### ISSUE TITLE
macOS Vircadia app crash on Quit

### ISSUE THREAD

**Issue by Steve-Pruneau** [vircadia/vircadia#998]
The macOS Vircadia interface app will crash and throw an error report when the user quits the app by doing either of these:

- Navigate to this menu option: Vircadia > Quit Vircadia
- Click the red "x" in the upper left of the window

This crash happens nearly every time the user quits the app.

[macOS_Vircadia_app_crash_report_2021-02-02.txt](https://github.com/vircadia/vircadia/files/5920212/macOS_Vircadia_app_crash_report_2021-02-02.txt)

**Comment by Steve-Pruneau** [comment]
With help from Moto, I ran an LLDB debugger back trace. [https://lldb.llvm.org]

Results attached.
[macOS_Viracadia_app_crash_LLDB_backtrace_2021-02-03.txt](https://github.com/vircadia/vircadia/files/5921837/macOS_Viracadia_app_crash_LLDB_backtrace_2021-02-03.txt)

**Comment by yuriymacdev** [comment]
Looking into this

**Comment by yuriymacdev** [comment]
A quick update on the findings.
As the crashlog states, this particular crash is happening because of accessing deallocated MainWindow object during teardown of one of the js-scripts.
I tried to make sure the window/menu objects are accessed only if a window is still shown on the screen, however that caused crashes in other places, also induced by accessing deallocated objects.

The root cause is the fact that some scripts are still running after multiple objects have already been deallocated during app termination.

A few questions:
- the same codebase does not crash on Windows and Linux, correct? Do you get any script exceptions in the logs when the app is closed on those systems?
- do you recall after which update it started to crash? (Qt version update, script editing?)
- is it crucial for the scripts to complete when the app is closing ?

**Comment by Steve-Pruneau** [comment]
Thank you for the update.

[_The same codebase does not crash on Windows and Linux, correct?_]
Correct for Windows 10. It does not crash on exit. `I believe also correct on Linux but will ask others.` For now, it is OK to assume the Vircadia app does not crash in Linux when the user Quits / Exits the app.

[_Do you recall after which update it started to crash? (Qt version update, script editing?)_]
I do not recall when it started crashing on macOS. That problem may have existed even when Vircadia took over from High Fidelity last year.

[_Is it crucial for the scripts to complete when the app is closing ?_]
I believe it is NOT required for scripts to complete. Scripts are created by people who create the virtual world, the same way a website developer creates web content. Scripts are not part of the Vircadia app itself. I think the way to handle scripts running in the Vircadia app should be similar to the way a browser would handle web site scripts when we want to Quit / Exit.
 `But I will ask Vircadia developers to give their comment on this.`

**Comment by digisomni** [comment]
Depending on the script, they should be able to fully shutdown if there are things they need to complete on "ending". I'd say think about it a bit like Windows. There are apps running and you could say force a restart.

So entity scripts maybe don't have to be waited on, but client scripts maybe should.

Maybe we should have a timer before forcefully killing them? Like start shutting down and if they don't all end by a certain time then it should log which ones did not kill and then just cut them? We should debug why some fail to end.

Entity scripts are in one script engine but client scripts are in their own, except for the default scripts (but those can be separated too using the developer menu). So we can debug which ones are problem children and find out what is going on.

**Comment by JulianGro** [comment]
Here a log off what a Linux stable build shutting down: 
[vircadia-log.txt](https://github.com/vircadia/vircadia/files/5954172/vircadia-log.txt)

**Comment by HifiExperiments** [comment]
this has been a longstanding issue.  essentially, the problem is that we DO need scripts to shutdown cleanly, so that script logic can complete.  but if a script calls a scripting API that marshals calls to another thread (via invoke or blocking invoke calls), we encounter issues on shutdown because we are in the process of destroying everything.

there is relevant mac-specific code in ScriptEngine::waitTillDoneRunning and at the bottom of Application::~Application()

also we actually do already have a "timer before forcefully killing them" in ScriptEngine::waitTillDoneRunning

**Comment by ctrlaltdavid** [comment]
@Steve-Pruneau 
- Does the Mac crash on exist happen for both official release builds and PR builds?
- Does the Mac crash on exit happen in the OSX QT 5.15.2 PR build? https://github.com/vircadia/vircadia/pull/968

**Comment by Steve-Pruneau** [comment]
[Does the Mac crash on exit happen for both official release builds and PR builds?]
Yes. It happens on official release builds and PR builds.

[Does the Mac crash on exit happen in the OSX QT 5.15.2 PR build? #968]
Yes. The crash on quite also happens with PR #968

**Comment by ctrlaltdavid** [comment]
1) Now that we're on Qt 5.15.2, I thought I'd try two tings just in case:
- Disable OSX-specific shut-down code in favour or using the same code as Windows & Linux. https://github.com/vircadia/vircadia/pull/1023
- Re-enable OSX-specific  shut-down code that was disabled in order to, ahem, fix shut-down crashes. https://github.com/vircadia/vircadia/pull/1023
I don't expect either of these to fix the OSX crash on exit problem, but they're worthwhile experiments to conduct.

2) The root cause of shut-down problems appears to be scripts trying to access C++ objects that have already been destroyed.
Perhaps shutdown should be reworked along the following lines or similar:
- Disable all user input. (Stop events continuing to come in to be processed.)
- Tell scripts to shut down. Terminate them if they haven't shut down in a timely manner.'
- And only after both the above have been done, proceed with shutting down the rest of the application.

**Comment by ctrlaltdavid** [comment]
Two further PRs to test just in case - Qt 5.15.2 versions of PRs 1023 and 1024:

- Disable OSX-specific shut-down code with Qt 5.15.2. https://github.com/vircadia/vircadia/pull/1031
- Re-enable disabled OSX-specific shut-down code with Qt 5.15.2. https://github.com/vircadia/vircadia/pull/1032
-

**Comment by ArcadeFever** [comment]
I think we have a winner with PR 1031.  With PR 1032 I was able to get it to crash consistently upon exit after using create tool.
